<!DOCTYPE html>
<html lang="" xml:lang="">
<head>

  <meta charset="utf-8" />
  <meta http-equiv="X-UA-Compatible" content="IE=edge" />
  <title>Chapter 3 An introduction to linkage mapping | Analysis of molecular data from fisheries and aquaculture</title>
  <meta name="description" content="Chapter 3 An introduction to linkage mapping | Analysis of molecular data from fisheries and aquaculture" />
  <meta name="generator" content="bookdown 0.21 and GitBook 2.6.7" />

  <meta property="og:title" content="Chapter 3 An introduction to linkage mapping | Analysis of molecular data from fisheries and aquaculture" />
  <meta property="og:type" content="book" />
  
  
  
  

  <meta name="twitter:card" content="summary" />
  <meta name="twitter:title" content="Chapter 3 An introduction to linkage mapping | Analysis of molecular data from fisheries and aquaculture" />
  
  
  

<meta name="author" content="PURA acuicultura" />


<meta name="date" content="2021-01-11" />

  <meta name="viewport" content="width=device-width, initial-scale=1" />
  <meta name="apple-mobile-web-app-capable" content="yes" />
  <meta name="apple-mobile-web-app-status-bar-style" content="black" />
  
  
<link rel="prev" href="differentiation-of-populations-using-pca.html"/>

<script src="libs/jquery-2.2.3/jquery.min.js"></script>
<link href="libs/gitbook-2.6.7/css/style.css" rel="stylesheet" />
<link href="libs/gitbook-2.6.7/css/plugin-table.css" rel="stylesheet" />
<link href="libs/gitbook-2.6.7/css/plugin-bookdown.css" rel="stylesheet" />
<link href="libs/gitbook-2.6.7/css/plugin-highlight.css" rel="stylesheet" />
<link href="libs/gitbook-2.6.7/css/plugin-search.css" rel="stylesheet" />
<link href="libs/gitbook-2.6.7/css/plugin-fontsettings.css" rel="stylesheet" />
<link href="libs/gitbook-2.6.7/css/plugin-clipboard.css" rel="stylesheet" />









<script src="libs/accessible-code-block-0.0.1/empty-anchor.js"></script>


<style type="text/css">
code.sourceCode > span { display: inline-block; line-height: 1.25; }
code.sourceCode > span { color: inherit; text-decoration: inherit; }
code.sourceCode > span:empty { height: 1.2em; }
.sourceCode { overflow: visible; }
code.sourceCode { white-space: pre; position: relative; }
pre.sourceCode { margin: 0; }
@media screen {
div.sourceCode { overflow: auto; }
}
@media print {
code.sourceCode { white-space: pre-wrap; }
code.sourceCode > span { text-indent: -5em; padding-left: 5em; }
}
pre.numberSource code
  { counter-reset: source-line 0; }
pre.numberSource code > span
  { position: relative; left: -4em; counter-increment: source-line; }
pre.numberSource code > span > a:first-child::before
  { content: counter(source-line);
    position: relative; left: -1em; text-align: right; vertical-align: baseline;
    border: none; display: inline-block;
    -webkit-touch-callout: none; -webkit-user-select: none;
    -khtml-user-select: none; -moz-user-select: none;
    -ms-user-select: none; user-select: none;
    padding: 0 4px; width: 4em;
    color: #aaaaaa;
  }
pre.numberSource { margin-left: 3em; border-left: 1px solid #aaaaaa;  padding-left: 4px; }
div.sourceCode
  {   }
@media screen {
code.sourceCode > span > a:first-child::before { text-decoration: underline; }
}
code span.al { color: #ff0000; font-weight: bold; } /* Alert */
code span.an { color: #60a0b0; font-weight: bold; font-style: italic; } /* Annotation */
code span.at { color: #7d9029; } /* Attribute */
code span.bn { color: #40a070; } /* BaseN */
code span.bu { } /* BuiltIn */
code span.cf { color: #007020; font-weight: bold; } /* ControlFlow */
code span.ch { color: #4070a0; } /* Char */
code span.cn { color: #880000; } /* Constant */
code span.co { color: #60a0b0; font-style: italic; } /* Comment */
code span.cv { color: #60a0b0; font-weight: bold; font-style: italic; } /* CommentVar */
code span.do { color: #ba2121; font-style: italic; } /* Documentation */
code span.dt { color: #902000; } /* DataType */
code span.dv { color: #40a070; } /* DecVal */
code span.er { color: #ff0000; font-weight: bold; } /* Error */
code span.ex { } /* Extension */
code span.fl { color: #40a070; } /* Float */
code span.fu { color: #06287e; } /* Function */
code span.im { } /* Import */
code span.in { color: #60a0b0; font-weight: bold; font-style: italic; } /* Information */
code span.kw { color: #007020; font-weight: bold; } /* Keyword */
code span.op { color: #666666; } /* Operator */
code span.ot { color: #007020; } /* Other */
code span.pp { color: #bc7a00; } /* Preprocessor */
code span.sc { color: #4070a0; } /* SpecialChar */
code span.ss { color: #bb6688; } /* SpecialString */
code span.st { color: #4070a0; } /* String */
code span.va { color: #19177c; } /* Variable */
code span.vs { color: #4070a0; } /* VerbatimString */
code span.wa { color: #60a0b0; font-weight: bold; font-style: italic; } /* Warning */
</style>

<link rel="stylesheet" href="style.css" type="text/css" />
</head>

<body>



  <div class="book without-animation with-summary font-size-2 font-family-1" data-basepath=".">

    <div class="book-summary">
      <nav role="navigation">

<ul class="summary">
<li><a href="./">Molecular methods</a></li>

<li class="divider"></li>
<li class="chapter" data-level="1" data-path="index.html"><a href="index.html"><i class="fa fa-check"></i><b>1</b> Introduction</a></li>
<li class="chapter" data-level="2" data-path="differentiation-of-populations-using-pca.html"><a href="differentiation-of-populations-using-pca.html"><i class="fa fa-check"></i><b>2</b> Differentiation of populations using PCA</a><ul>
<li class="chapter" data-level="2.1" data-path="differentiation-of-populations-using-pca.html"><a href="differentiation-of-populations-using-pca.html#introduction-1"><i class="fa fa-check"></i><b>2.1</b> Introduction</a></li>
<li class="chapter" data-level="2.2" data-path="differentiation-of-populations-using-pca.html"><a href="differentiation-of-populations-using-pca.html#background-on-the-fish-populations"><i class="fa fa-check"></i><b>2.2</b> Background on the fish populations</a></li>
<li class="chapter" data-level="2.3" data-path="differentiation-of-populations-using-pca.html"><a href="differentiation-of-populations-using-pca.html#snp-results"><i class="fa fa-check"></i><b>2.3</b> SNP results</a></li>
<li class="chapter" data-level="2.4" data-path="differentiation-of-populations-using-pca.html"><a href="differentiation-of-populations-using-pca.html#pca-analysis"><i class="fa fa-check"></i><b>2.4</b> PCA analysis</a></li>
<li class="chapter" data-level="2.5" data-path="differentiation-of-populations-using-pca.html"><a href="differentiation-of-populations-using-pca.html#assessment-of-the-principle-components"><i class="fa fa-check"></i><b>2.5</b> Assessment of the principle components</a></li>
<li class="chapter" data-level="2.6" data-path="differentiation-of-populations-using-pca.html"><a href="differentiation-of-populations-using-pca.html#pca-graphical-analysis"><i class="fa fa-check"></i><b>2.6</b> PCA graphical analysis</a></li>
<li class="chapter" data-level="2.7" data-path="differentiation-of-populations-using-pca.html"><a href="differentiation-of-populations-using-pca.html#identify-significant-snp"><i class="fa fa-check"></i><b>2.7</b> Identify significant SNP</a></li>
</ul></li>
<li class="chapter" data-level="3" data-path="an-introduction-to-linkage-mapping.html"><a href="an-introduction-to-linkage-mapping.html"><i class="fa fa-check"></i><b>3</b> An introduction to linkage mapping</a><ul>
<li class="chapter" data-level="3.1" data-path="an-introduction-to-linkage-mapping.html"><a href="an-introduction-to-linkage-mapping.html#creating-the-parental-haplotypes"><i class="fa fa-check"></i><b>3.1</b> Creating the parental haplotypes</a></li>
</ul></li>
<li class="divider"></li>
<li><a href="https://github.com/rstudio/bookdown" target="blank">Published with bookdown</a></li>
</ul>

      </nav>
    </div>

    <div class="book-body">
      <div class="body-inner">
        <div class="book-header" role="navigation">
          <h1>
            <i class="fa fa-circle-o-notch fa-spin"></i><a href="./">Analysis of molecular data from fisheries and aquaculture</a>
          </h1>
        </div>

        <div class="page-wrapper" tabindex="-1" role="main">
          <div class="page-inner">

            <section class="normal" id="section-">
<div id="an-introduction-to-linkage-mapping" class="section level1">
<h1><span class="header-section-number">Chapter 3</span> An introduction to linkage mapping</h1>
<p>Here we create a simple simulator to demonstrate the underlying principles of linkage mapping.</p>
<p>Simulations are conducted using <code>AlphaSimR</code>. Analysis is carried out using Tidyverse tools such as <code>dplyr</code> and <code>ggplot</code>. The simulator is constructed with three independent variables, chromosome number, segregation site number, and offspring number as a foundation for understanding how these variables contribute to the precision and accuracy of the linkage maps themselves.</p>
<div class="sourceCode" id="cb9"><pre class="sourceCode r"><code class="sourceCode r"><span id="cb9-1"><a href="an-introduction-to-linkage-mapping.html#cb9-1"></a><span class="kw">library</span>(AlphaSimR)</span>
<span id="cb9-2"><a href="an-introduction-to-linkage-mapping.html#cb9-2"></a><span class="kw">library</span>(tidyverse)</span>
<span id="cb9-3"><a href="an-introduction-to-linkage-mapping.html#cb9-3"></a></span>
<span id="cb9-4"><a href="an-introduction-to-linkage-mapping.html#cb9-4"></a>n_chr &lt;-<span class="st"> </span><span class="dv">1</span> <span class="co"># variable 1, number of chromosomes in genome</span></span>
<span id="cb9-5"><a href="an-introduction-to-linkage-mapping.html#cb9-5"></a>n_seg &lt;-<span class="st"> </span><span class="dv">10</span> <span class="co"># variable 2, number of segregation sites per chromosome</span></span>
<span id="cb9-6"><a href="an-introduction-to-linkage-mapping.html#cb9-6"></a>n_off &lt;-<span class="st"> </span><span class="dv">100</span> <span class="co"># variable 3. number of offspring per parental cross</span></span></code></pre></div>
<p>The selection of parents is the most critical aspect of successful linkage mapping. Parents are not selected based on specific traits or physical health but on characteristics of their genomes that facilitate mapping. We will start with the most ‘textbook’ of approaches, crossing of a completely heterozygous individual (10) with a completely homozygous recessive individual (00). under such circumstances, offspring are expected to be 50% heterozygous (10) and 50% homozygous recessive (00). As such, any deviation from these percentages provides evidence for linkage between segregation sites, and the data needed to create the linkage map itself.</p>
<div id="creating-the-parental-haplotypes" class="section level2">
<h2><span class="header-section-number">3.1</span> Creating the parental haplotypes</h2>
<p>Rather than using <code>quickHaplo()</code>, custom haplotypes will be defined using the <code>newMapPop()</code> function. As arguments, <code>newMapPop()</code> requires a genetic map and a list of corresponding haplotypes. The following code first creates the genetic map. The <code>seq()</code> function is used to create a numerical list from 0 to 1. By setting <code>length.out</code> equal to the <code>n_seg</code>, a list of equally sized segregation sites is created. It should be noted that defining the genetic map from 0 to 1 results in a chromosome length of 1 Morgan (NB: 100 cM).</p>
<div class="sourceCode" id="cb10"><pre class="sourceCode r"><code class="sourceCode r"><span id="cb10-1"><a href="an-introduction-to-linkage-mapping.html#cb10-1"></a>genMap &lt;-<span class="st"> </span><span class="kw">list</span>(<span class="kw">seq</span>(<span class="dv">0</span>, <span class="dv">1</span>, <span class="dt">length.out =</span> n_seg)) <span class="co"># list of segregation sites</span></span>
<span id="cb10-2"><a href="an-introduction-to-linkage-mapping.html#cb10-2"></a><span class="kw">round</span>(<span class="kw">unlist</span>(genMap[<span class="dv">1</span>]),<span class="dv">2</span>) <span class="co"># show the list of segregation sites</span></span></code></pre></div>
<pre><code>##  [1] 0.00 0.11 0.22 0.33 0.44 0.56 0.67 0.78 0.89 1.00</code></pre>
<div class="sourceCode" id="cb12"><pre class="sourceCode r"><code class="sourceCode r"><span id="cb12-1"><a href="an-introduction-to-linkage-mapping.html#cb12-1"></a>chr1 &lt;-<span class="st"> </span><span class="kw">c</span>(<span class="kw">rep</span>(<span class="dv">1</span>, n_seg), <span class="co"># haplotype 1 of heterozygote parent</span></span>
<span id="cb12-2"><a href="an-introduction-to-linkage-mapping.html#cb12-2"></a>          <span class="kw">rep</span>(<span class="dv">0</span>, n_seg), <span class="co"># haplotype 2 of heterozygote parent</span></span>
<span id="cb12-3"><a href="an-introduction-to-linkage-mapping.html#cb12-3"></a>          <span class="kw">rep</span>(<span class="dv">0</span>, n_seg), <span class="co"># haplotypes 1 of homozygous parent</span></span>
<span id="cb12-4"><a href="an-introduction-to-linkage-mapping.html#cb12-4"></a>          <span class="kw">rep</span>(<span class="dv">0</span>, n_seg)) <span class="co"># haplotypes 2 of homozygous parent</span></span>
<span id="cb12-5"><a href="an-introduction-to-linkage-mapping.html#cb12-5"></a></span>
<span id="cb12-6"><a href="an-introduction-to-linkage-mapping.html#cb12-6"></a>chr1 &lt;-<span class="st"> </span><span class="kw">matrix</span>(chr1, <span class="dt">nrow =</span> <span class="dv">4</span>, <span class="dt">ncol =</span> n_seg, <span class="dt">byrow =</span> <span class="ot">TRUE</span>) <span class="co"># coerce into a matrix</span></span>
<span id="cb12-7"><a href="an-introduction-to-linkage-mapping.html#cb12-7"></a>chr1 <span class="co"># show the matrix of haplotypes</span></span></code></pre></div>
<pre><code>##      [,1] [,2] [,3] [,4] [,5] [,6] [,7] [,8] [,9] [,10]
## [1,]    1    1    1    1    1    1    1    1    1     1
## [2,]    0    0    0    0    0    0    0    0    0     0
## [3,]    0    0    0    0    0    0    0    0    0     0
## [4,]    0    0    0    0    0    0    0    0    0     0</code></pre>
<div class="sourceCode" id="cb14"><pre class="sourceCode r"><code class="sourceCode r"><span id="cb14-1"><a href="an-introduction-to-linkage-mapping.html#cb14-1"></a>pop_haplos &lt;-<span class="st"> </span><span class="kw">newMapPop</span>(<span class="dt">genMap =</span> genMap,</span>
<span id="cb14-2"><a href="an-introduction-to-linkage-mapping.html#cb14-2"></a>                        <span class="dt">haplotypes =</span> <span class="kw">list</span>(chr1))</span></code></pre></div>

</div>
</div>
            </section>

          </div>
        </div>
      </div>
<a href="differentiation-of-populations-using-pca.html" class="navigation navigation-prev navigation-unique" aria-label="Previous page"><i class="fa fa-angle-left"></i></a>

    </div>
  </div>
<script src="libs/gitbook-2.6.7/js/app.min.js"></script>
<script src="libs/gitbook-2.6.7/js/lunr.js"></script>
<script src="libs/gitbook-2.6.7/js/clipboard.min.js"></script>
<script src="libs/gitbook-2.6.7/js/plugin-search.js"></script>
<script src="libs/gitbook-2.6.7/js/plugin-sharing.js"></script>
<script src="libs/gitbook-2.6.7/js/plugin-fontsettings.js"></script>
<script src="libs/gitbook-2.6.7/js/plugin-bookdown.js"></script>
<script src="libs/gitbook-2.6.7/js/jquery.highlight.js"></script>
<script src="libs/gitbook-2.6.7/js/plugin-clipboard.js"></script>
<script>
gitbook.require(["gitbook"], function(gitbook) {
gitbook.start({
"sharing": {
"github": false,
"facebook": true,
"twitter": true,
"linkedin": false,
"weibo": false,
"instapaper": false,
"vk": false,
"all": ["facebook", "twitter", "linkedin", "weibo", "instapaper"]
},
"fontsettings": {
"theme": "white",
"family": "sans",
"size": 2
},
"edit": {
"link": "https://github.com/pura-acuicultura/Molecular-methods/edit/master/03-linkage.Rmd",
"text": "Edit"
},
"history": {
"link": null,
"text": null
},
"view": {
"link": "https://github.com/pura-acuicultura/Molecular-methods/blob/master/03-linkage.Rmd",
"text": null
},
"download": null,
"toc": {
"collapse": "subsection"
}
});
});
</script>

</body>

</html>
