<!DOCTYPE html>
<html lang="" xml:lang="">
<head>

  <meta charset="utf-8" />
  <meta http-equiv="X-UA-Compatible" content="IE=edge" />
  <title>Chapter 2 Differentiation of populations using PCA | Analysis of molecular data from fisheries and aquaculture</title>
  <meta name="description" content="Chapter 2 Differentiation of populations using PCA | Analysis of molecular data from fisheries and aquaculture" />
  <meta name="generator" content="bookdown 0.21 and GitBook 2.6.7" />

  <meta property="og:title" content="Chapter 2 Differentiation of populations using PCA | Analysis of molecular data from fisheries and aquaculture" />
  <meta property="og:type" content="book" />
  
  
  
  

  <meta name="twitter:card" content="summary" />
  <meta name="twitter:title" content="Chapter 2 Differentiation of populations using PCA | Analysis of molecular data from fisheries and aquaculture" />
  
  
  

<meta name="author" content="PURA acuicultura" />


<meta name="date" content="2021-01-11" />

  <meta name="viewport" content="width=device-width, initial-scale=1" />
  <meta name="apple-mobile-web-app-capable" content="yes" />
  <meta name="apple-mobile-web-app-status-bar-style" content="black" />
  
  
<link rel="prev" href="index.html"/>
<link rel="next" href="an-introduction-to-linkage-mapping.html"/>
<script src="libs/jquery-2.2.3/jquery.min.js"></script>
<link href="libs/gitbook-2.6.7/css/style.css" rel="stylesheet" />
<link href="libs/gitbook-2.6.7/css/plugin-table.css" rel="stylesheet" />
<link href="libs/gitbook-2.6.7/css/plugin-bookdown.css" rel="stylesheet" />
<link href="libs/gitbook-2.6.7/css/plugin-highlight.css" rel="stylesheet" />
<link href="libs/gitbook-2.6.7/css/plugin-search.css" rel="stylesheet" />
<link href="libs/gitbook-2.6.7/css/plugin-fontsettings.css" rel="stylesheet" />
<link href="libs/gitbook-2.6.7/css/plugin-clipboard.css" rel="stylesheet" />









<script src="libs/accessible-code-block-0.0.1/empty-anchor.js"></script>


<style type="text/css">
code.sourceCode > span { display: inline-block; line-height: 1.25; }
code.sourceCode > span { color: inherit; text-decoration: inherit; }
code.sourceCode > span:empty { height: 1.2em; }
.sourceCode { overflow: visible; }
code.sourceCode { white-space: pre; position: relative; }
pre.sourceCode { margin: 0; }
@media screen {
div.sourceCode { overflow: auto; }
}
@media print {
code.sourceCode { white-space: pre-wrap; }
code.sourceCode > span { text-indent: -5em; padding-left: 5em; }
}
pre.numberSource code
  { counter-reset: source-line 0; }
pre.numberSource code > span
  { position: relative; left: -4em; counter-increment: source-line; }
pre.numberSource code > span > a:first-child::before
  { content: counter(source-line);
    position: relative; left: -1em; text-align: right; vertical-align: baseline;
    border: none; display: inline-block;
    -webkit-touch-callout: none; -webkit-user-select: none;
    -khtml-user-select: none; -moz-user-select: none;
    -ms-user-select: none; user-select: none;
    padding: 0 4px; width: 4em;
    color: #aaaaaa;
  }
pre.numberSource { margin-left: 3em; border-left: 1px solid #aaaaaa;  padding-left: 4px; }
div.sourceCode
  {   }
@media screen {
code.sourceCode > span > a:first-child::before { text-decoration: underline; }
}
code span.al { color: #ff0000; font-weight: bold; } /* Alert */
code span.an { color: #60a0b0; font-weight: bold; font-style: italic; } /* Annotation */
code span.at { color: #7d9029; } /* Attribute */
code span.bn { color: #40a070; } /* BaseN */
code span.bu { } /* BuiltIn */
code span.cf { color: #007020; font-weight: bold; } /* ControlFlow */
code span.ch { color: #4070a0; } /* Char */
code span.cn { color: #880000; } /* Constant */
code span.co { color: #60a0b0; font-style: italic; } /* Comment */
code span.cv { color: #60a0b0; font-weight: bold; font-style: italic; } /* CommentVar */
code span.do { color: #ba2121; font-style: italic; } /* Documentation */
code span.dt { color: #902000; } /* DataType */
code span.dv { color: #40a070; } /* DecVal */
code span.er { color: #ff0000; font-weight: bold; } /* Error */
code span.ex { } /* Extension */
code span.fl { color: #40a070; } /* Float */
code span.fu { color: #06287e; } /* Function */
code span.im { } /* Import */
code span.in { color: #60a0b0; font-weight: bold; font-style: italic; } /* Information */
code span.kw { color: #007020; font-weight: bold; } /* Keyword */
code span.op { color: #666666; } /* Operator */
code span.ot { color: #007020; } /* Other */
code span.pp { color: #bc7a00; } /* Preprocessor */
code span.sc { color: #4070a0; } /* SpecialChar */
code span.ss { color: #bb6688; } /* SpecialString */
code span.st { color: #4070a0; } /* String */
code span.va { color: #19177c; } /* Variable */
code span.vs { color: #4070a0; } /* VerbatimString */
code span.wa { color: #60a0b0; font-weight: bold; font-style: italic; } /* Warning */
</style>

<link rel="stylesheet" href="style.css" type="text/css" />
</head>

<body>



  <div class="book without-animation with-summary font-size-2 font-family-1" data-basepath=".">

    <div class="book-summary">
      <nav role="navigation">

<ul class="summary">
<li><a href="./">Molecular methods</a></li>

<li class="divider"></li>
<li class="chapter" data-level="1" data-path="index.html"><a href="index.html"><i class="fa fa-check"></i><b>1</b> Introduction</a></li>
<li class="chapter" data-level="2" data-path="differentiation-of-populations-using-pca.html"><a href="differentiation-of-populations-using-pca.html"><i class="fa fa-check"></i><b>2</b> Differentiation of populations using PCA</a><ul>
<li class="chapter" data-level="2.1" data-path="differentiation-of-populations-using-pca.html"><a href="differentiation-of-populations-using-pca.html#introduction-1"><i class="fa fa-check"></i><b>2.1</b> Introduction</a></li>
<li class="chapter" data-level="2.2" data-path="differentiation-of-populations-using-pca.html"><a href="differentiation-of-populations-using-pca.html#background-on-the-fish-populations"><i class="fa fa-check"></i><b>2.2</b> Background on the fish populations</a></li>
<li class="chapter" data-level="2.3" data-path="differentiation-of-populations-using-pca.html"><a href="differentiation-of-populations-using-pca.html#snp-results"><i class="fa fa-check"></i><b>2.3</b> SNP results</a></li>
<li class="chapter" data-level="2.4" data-path="differentiation-of-populations-using-pca.html"><a href="differentiation-of-populations-using-pca.html#pca-analysis"><i class="fa fa-check"></i><b>2.4</b> PCA analysis</a></li>
<li class="chapter" data-level="2.5" data-path="differentiation-of-populations-using-pca.html"><a href="differentiation-of-populations-using-pca.html#assessment-of-the-principle-components"><i class="fa fa-check"></i><b>2.5</b> Assessment of the principle components</a></li>
<li class="chapter" data-level="2.6" data-path="differentiation-of-populations-using-pca.html"><a href="differentiation-of-populations-using-pca.html#pca-graphical-analysis"><i class="fa fa-check"></i><b>2.6</b> PCA graphical analysis</a></li>
<li class="chapter" data-level="2.7" data-path="differentiation-of-populations-using-pca.html"><a href="differentiation-of-populations-using-pca.html#identify-significant-snp"><i class="fa fa-check"></i><b>2.7</b> Identify significant SNP</a></li>
</ul></li>
<li class="chapter" data-level="3" data-path="an-introduction-to-linkage-mapping.html"><a href="an-introduction-to-linkage-mapping.html"><i class="fa fa-check"></i><b>3</b> An introduction to linkage mapping</a><ul>
<li class="chapter" data-level="3.1" data-path="an-introduction-to-linkage-mapping.html"><a href="an-introduction-to-linkage-mapping.html#creating-the-parental-haplotypes"><i class="fa fa-check"></i><b>3.1</b> Creating the parental haplotypes</a></li>
</ul></li>
<li class="divider"></li>
<li><a href="https://github.com/rstudio/bookdown" target="blank">Published with bookdown</a></li>
</ul>

      </nav>
    </div>

    <div class="book-body">
      <div class="body-inner">
        <div class="book-header" role="navigation">
          <h1>
            <i class="fa fa-circle-o-notch fa-spin"></i><a href="./">Analysis of molecular data from fisheries and aquaculture</a>
          </h1>
        </div>

        <div class="page-wrapper" tabindex="-1" role="main">
          <div class="page-inner">

            <section class="normal" id="section-">
<div id="differentiation-of-populations-using-pca" class="section level1">
<h1><span class="header-section-number">Chapter 2</span> Differentiation of populations using PCA</h1>
<div id="introduction-1" class="section level2">
<h2><span class="header-section-number">2.1</span> Introduction</h2>
<p>Principle component analysis (PCA) is one of several methods for differentiating populations through molecular markers. At its most basic level it is a rapid an easy analysis to perform. Here, individuals from two fish populations are assessed using a low density, single nucleotide polymorphism (SNP) panel. PCA is explored in the context of developing methods for classifying individuals based only on molecular analysis.</p>
<div class="sourceCode" id="cb1"><pre class="sourceCode r"><code class="sourceCode r"><span id="cb1-1"><a href="differentiation-of-populations-using-pca.html#cb1-1"></a><span class="kw">library</span>(AlphaSimR)</span>
<span id="cb1-2"><a href="differentiation-of-populations-using-pca.html#cb1-2"></a><span class="kw">library</span>(tidyverse)</span>
<span id="cb1-3"><a href="differentiation-of-populations-using-pca.html#cb1-3"></a><span class="kw">library</span>(knitr)</span></code></pre></div>
</div>
<div id="background-on-the-fish-populations" class="section level2">
<h2><span class="header-section-number">2.2</span> Background on the fish populations</h2>
<p>A population of haplotypes is first made
From the haplotypes, two populations are formed, A and B
The populations are managed at two different hatcheries
Hatchery A maintains a large effective population size throughout 10 generations
Hatchery B, a much smaller effective population size also for 10 generations
Inbreeding and genetic drift increase the homozygocity of population B
We will use principle component analysis to try and differentiate the populations at each generation</p>
<div class="sourceCode" id="cb2"><pre class="sourceCode r"><code class="sourceCode r"><span id="cb2-1"><a href="differentiation-of-populations-using-pca.html#cb2-1"></a>pop_haplos &lt;-<span class="st"> </span><span class="kw">quickHaplo</span>(<span class="dt">nInd =</span> <span class="dv">10</span>, <span class="co"># number of individuals</span></span>
<span id="cb2-2"><a href="differentiation-of-populations-using-pca.html#cb2-2"></a>                         <span class="dt">nChr =</span> <span class="dv">3</span>, <span class="co"># number of chromosomes</span></span>
<span id="cb2-3"><a href="differentiation-of-populations-using-pca.html#cb2-3"></a>                         <span class="dt">segSites =</span> <span class="dv">1000</span>, <span class="co"># number of segregation sites</span></span>
<span id="cb2-4"><a href="differentiation-of-populations-using-pca.html#cb2-4"></a>                         <span class="dt">genLen =</span> <span class="dv">1</span>, <span class="co"># genetic length of chromosomes</span></span>
<span id="cb2-5"><a href="differentiation-of-populations-using-pca.html#cb2-5"></a>                         <span class="dt">ploidy =</span> 2L, <span class="co"># ploidy level of organism</span></span>
<span id="cb2-6"><a href="differentiation-of-populations-using-pca.html#cb2-6"></a>                         <span class="dt">inbred =</span> <span class="ot">TRUE</span>) <span class="co"># are founders inbred?</span></span>
<span id="cb2-7"><a href="differentiation-of-populations-using-pca.html#cb2-7"></a></span>
<span id="cb2-8"><a href="differentiation-of-populations-using-pca.html#cb2-8"></a>SP =<span class="st"> </span>SimParam<span class="op">$</span><span class="kw">new</span>(pop_haplos) <span class="co"># create a variable for new simulation parameters</span></span>
<span id="cb2-9"><a href="differentiation-of-populations-using-pca.html#cb2-9"></a></span>
<span id="cb2-10"><a href="differentiation-of-populations-using-pca.html#cb2-10"></a>SP<span class="op">$</span><span class="kw">addTraitA</span>(<span class="dt">nQtlPerChr =</span> <span class="dv">100</span>, <span class="co"># number of QTL for trait</span></span>
<span id="cb2-11"><a href="differentiation-of-populations-using-pca.html#cb2-11"></a>             <span class="dt">mean =</span> <span class="dv">500</span>, <span class="co"># mean genetic value of the trait</span></span>
<span id="cb2-12"><a href="differentiation-of-populations-using-pca.html#cb2-12"></a>             <span class="dt">var =</span> <span class="dv">100</span>, <span class="co"># variance of trait</span></span>
<span id="cb2-13"><a href="differentiation-of-populations-using-pca.html#cb2-13"></a>             <span class="dt">corA =</span> <span class="ot">NULL</span>, <span class="co"># matrix of correlations between additive effects</span></span>
<span id="cb2-14"><a href="differentiation-of-populations-using-pca.html#cb2-14"></a>             <span class="dt">gamma =</span> <span class="ot">FALSE</span>, <span class="co"># to use a gamma distribution in place of a normal distribution</span></span>
<span id="cb2-15"><a href="differentiation-of-populations-using-pca.html#cb2-15"></a>             <span class="dt">shape =</span> <span class="dv">1</span>, <span class="co"># shape parameter for gamma distribution only</span></span>
<span id="cb2-16"><a href="differentiation-of-populations-using-pca.html#cb2-16"></a>             <span class="dt">force =</span> <span class="ot">FALSE</span>) <span class="co"># keep false until this is understood!</span></span>
<span id="cb2-17"><a href="differentiation-of-populations-using-pca.html#cb2-17"></a></span>
<span id="cb2-18"><a href="differentiation-of-populations-using-pca.html#cb2-18"></a>SP<span class="op">$</span><span class="kw">setVarE</span>(<span class="dt">h2 =</span> <span class="fl">0.3</span>, <span class="co"># vector of narrow sense heritability</span></span>
<span id="cb2-19"><a href="differentiation-of-populations-using-pca.html#cb2-19"></a>           <span class="dt">H2 =</span> <span class="ot">NULL</span>, <span class="co"># vector of broad sense heritability</span></span>
<span id="cb2-20"><a href="differentiation-of-populations-using-pca.html#cb2-20"></a>           <span class="dt">varE =</span> <span class="ot">NULL</span>) <span class="co"># vector of error variances</span></span>
<span id="cb2-21"><a href="differentiation-of-populations-using-pca.html#cb2-21"></a></span>
<span id="cb2-22"><a href="differentiation-of-populations-using-pca.html#cb2-22"></a>SP<span class="op">$</span><span class="kw">addSnpChip</span>(<span class="dt">nSnpPerChr =</span> <span class="dv">32</span>, <span class="co"># number of SNPs per chromosome</span></span>
<span id="cb2-23"><a href="differentiation-of-populations-using-pca.html#cb2-23"></a>              <span class="dt">minSnpFreq =</span> <span class="ot">NULL</span>, <span class="co"># minimum allowable frequency for SNP loci</span></span>
<span id="cb2-24"><a href="differentiation-of-populations-using-pca.html#cb2-24"></a>              <span class="dt">refPop =</span> <span class="ot">NULL</span>) <span class="co"># reference population for calculating SNP frequency</span></span></code></pre></div>
<div class="sourceCode" id="cb3"><pre class="sourceCode r"><code class="sourceCode r"><span id="cb3-1"><a href="differentiation-of-populations-using-pca.html#cb3-1"></a>gen &lt;-<span class="st"> </span><span class="dv">10</span> <span class="co"># specify the number of generations for which selection will be carried out</span></span>
<span id="cb3-2"><a href="differentiation-of-populations-using-pca.html#cb3-2"></a></span>
<span id="cb3-3"><a href="differentiation-of-populations-using-pca.html#cb3-3"></a><span class="co">#__ 3.1 Develop population A #####</span></span>
<span id="cb3-4"><a href="differentiation-of-populations-using-pca.html#cb3-4"></a>pop_A &lt;-<span class="st"> </span><span class="kw">newPop</span>(<span class="dt">rawPop =</span> pop_haplos, <span class="co"># object of MapPop-class</span></span>
<span id="cb3-5"><a href="differentiation-of-populations-using-pca.html#cb3-5"></a>                <span class="dt">mother =</span> <span class="ot">NULL</span>, <span class="co"># optional id for mothers</span></span>
<span id="cb3-6"><a href="differentiation-of-populations-using-pca.html#cb3-6"></a>                <span class="dt">father =</span> <span class="ot">NULL</span>, <span class="co"># optional id for fathers</span></span>
<span id="cb3-7"><a href="differentiation-of-populations-using-pca.html#cb3-7"></a>                <span class="dt">origM =</span> <span class="ot">NULL</span>, <span class="co"># optional alternative id for mothers</span></span>
<span id="cb3-8"><a href="differentiation-of-populations-using-pca.html#cb3-8"></a>                <span class="dt">origF =</span> <span class="ot">NULL</span>, <span class="co"># optional alternative id for fathers</span></span>
<span id="cb3-9"><a href="differentiation-of-populations-using-pca.html#cb3-9"></a>                <span class="dt">isDH =</span> <span class="ot">FALSE</span>, <span class="co"># indicate if double haploids and/or inbred</span></span>
<span id="cb3-10"><a href="differentiation-of-populations-using-pca.html#cb3-10"></a>                <span class="dt">simParam =</span> SP) <span class="co"># SimParam objects</span></span>
<span id="cb3-11"><a href="differentiation-of-populations-using-pca.html#cb3-11"></a></span>
<span id="cb3-12"><a href="differentiation-of-populations-using-pca.html#cb3-12"></a>POP_A &lt;-<span class="st"> </span><span class="kw">vector</span>(<span class="dt">length =</span> gen, <span class="dt">mode =</span> <span class="st">&quot;list&quot;</span>) <span class="co"># create a vector of lists to store the population for each generation</span></span>
<span id="cb3-13"><a href="differentiation-of-populations-using-pca.html#cb3-13"></a>POP_A[[<span class="dv">1</span>]] &lt;-<span class="st"> </span>pop_A <span class="co"># add the founder population to the vector of lists</span></span>
<span id="cb3-14"><a href="differentiation-of-populations-using-pca.html#cb3-14"></a></span>
<span id="cb3-15"><a href="differentiation-of-populations-using-pca.html#cb3-15"></a><span class="cf">for</span> (i <span class="cf">in</span> <span class="dv">2</span><span class="op">:</span>gen) {</span>
<span id="cb3-16"><a href="differentiation-of-populations-using-pca.html#cb3-16"></a>  POP_A[[i]] &lt;-<span class="st"> </span><span class="kw">randCross</span>(<span class="dt">pop =</span> POP_A[[i <span class="op">-</span><span class="st"> </span><span class="dv">1</span>]],</span>
<span id="cb3-17"><a href="differentiation-of-populations-using-pca.html#cb3-17"></a>                          <span class="dt">nCrosses =</span> <span class="dv">10</span>,</span>
<span id="cb3-18"><a href="differentiation-of-populations-using-pca.html#cb3-18"></a>                          <span class="dt">nProgeny =</span> <span class="dv">1</span>,</span>
<span id="cb3-19"><a href="differentiation-of-populations-using-pca.html#cb3-19"></a>                          <span class="dt">balance =</span> <span class="ot">TRUE</span>,</span>
<span id="cb3-20"><a href="differentiation-of-populations-using-pca.html#cb3-20"></a>                          <span class="dt">parents =</span> <span class="ot">NULL</span>,</span>
<span id="cb3-21"><a href="differentiation-of-populations-using-pca.html#cb3-21"></a>                          <span class="dt">simParam =</span> SP)}</span></code></pre></div>
</div>
<div id="snp-results" class="section level2">
<h2><span class="header-section-number">2.3</span> SNP results</h2>
<p>SNP results are returned as a data matrix of m rows and n columns. Each row is an individual fish and each column is a different SNP. 10 individuals from each hatchery were assessed so m = 20 observations exist. The low density SNP panel consists of 96 SNP so n = 96 variables exist. For each individual, each SNP is assessed as 0, 1, or 2 where 0 equates to homzygous recessive, 1 to heterozygous, and 2 to homozygous dominant. It is important to note that the terms recessive and dominant do not necessarily correspond with the genetic definitions of these terms. In the case of an SNP panel, they correspond with the allele-by-reporter dye combination chosen during assay development.</p>
<p>The first few rows and columns of the SNP results are shown. Note that the matrix contains only data and that fish IDs and SNP IDs exist only as row and column names. This is clear evidence that PCA analysis incorporates no prior knowledge of population origin and relies completely on genotypic data.</p>
<div class="sourceCode" id="cb4"><pre class="sourceCode r"><code class="sourceCode r"><span id="cb4-1"><a href="differentiation-of-populations-using-pca.html#cb4-1"></a>c_gen &lt;-<span class="st"> </span><span class="dv">10</span> <span class="co"># specify the generation at which PCA will be performed</span></span>
<span id="cb4-2"><a href="differentiation-of-populations-using-pca.html#cb4-2"></a></span>
<span id="cb4-3"><a href="differentiation-of-populations-using-pca.html#cb4-3"></a>snp_A &lt;-<span class="st"> </span><span class="kw">pullSnpGeno</span>(<span class="dt">pop =</span> POP_A[[c_gen]], <span class="co"># Pop-class object</span></span>
<span id="cb4-4"><a href="differentiation-of-populations-using-pca.html#cb4-4"></a>                     <span class="dt">snpChip =</span> <span class="dv">1</span>, <span class="co"># which chip to use</span></span>
<span id="cb4-5"><a href="differentiation-of-populations-using-pca.html#cb4-5"></a>                     <span class="dt">chr =</span> <span class="ot">NULL</span>, <span class="co"># chromosome number, NULL = all</span></span>
<span id="cb4-6"><a href="differentiation-of-populations-using-pca.html#cb4-6"></a>                     <span class="dt">simParam =</span> SP) <span class="co"># SimParam object</span></span>
<span id="cb4-7"><a href="differentiation-of-populations-using-pca.html#cb4-7"></a></span>
<span id="cb4-8"><a href="differentiation-of-populations-using-pca.html#cb4-8"></a>snp_B &lt;-<span class="st"> </span><span class="kw">pullSnpGeno</span>(<span class="dt">pop =</span> POP_B[[c_gen]], <span class="co"># Pop-class object</span></span>
<span id="cb4-9"><a href="differentiation-of-populations-using-pca.html#cb4-9"></a>                     <span class="dt">snpChip =</span> <span class="dv">1</span>, <span class="co"># which chip to use</span></span>
<span id="cb4-10"><a href="differentiation-of-populations-using-pca.html#cb4-10"></a>                     <span class="dt">chr =</span> <span class="ot">NULL</span>, <span class="co"># chromosome number, NULL = all</span></span>
<span id="cb4-11"><a href="differentiation-of-populations-using-pca.html#cb4-11"></a>                     <span class="dt">simParam =</span> SP) <span class="co"># SimParam object</span></span>
<span id="cb4-12"><a href="differentiation-of-populations-using-pca.html#cb4-12"></a></span>
<span id="cb4-13"><a href="differentiation-of-populations-using-pca.html#cb4-13"></a>snp_pops &lt;-<span class="st"> </span><span class="kw">rbind</span>(snp_A, snp_B)</span>
<span id="cb4-14"><a href="differentiation-of-populations-using-pca.html#cb4-14"></a><span class="kw">kable</span>(snp_pops[<span class="dv">1</span><span class="op">:</span><span class="dv">20</span>, <span class="dv">1</span><span class="op">:</span><span class="dv">12</span>])</span></code></pre></div>
<table>
<thead>
<tr class="header">
<th align="left"></th>
<th align="right">SNP_1</th>
<th align="right">SNP_2</th>
<th align="right">SNP_3</th>
<th align="right">SNP_4</th>
<th align="right">SNP_5</th>
<th align="right">SNP_6</th>
<th align="right">SNP_7</th>
<th align="right">SNP_8</th>
<th align="right">SNP_9</th>
<th align="right">SNP_10</th>
<th align="right">SNP_11</th>
<th align="right">SNP_12</th>
</tr>
</thead>
<tbody>
<tr class="odd">
<td align="left">91</td>
<td align="right">0</td>
<td align="right">1</td>
<td align="right">1</td>
<td align="right">0</td>
<td align="right">1</td>
<td align="right">1</td>
<td align="right">1</td>
<td align="right">2</td>
<td align="right">2</td>
<td align="right">2</td>
<td align="right">0</td>
<td align="right">0</td>
</tr>
<tr class="even">
<td align="left">92</td>
<td align="right">0</td>
<td align="right">1</td>
<td align="right">2</td>
<td align="right">0</td>
<td align="right">1</td>
<td align="right">1</td>
<td align="right">0</td>
<td align="right">2</td>
<td align="right">1</td>
<td align="right">1</td>
<td align="right">0</td>
<td align="right">0</td>
</tr>
<tr class="odd">
<td align="left">93</td>
<td align="right">0</td>
<td align="right">2</td>
<td align="right">2</td>
<td align="right">0</td>
<td align="right">0</td>
<td align="right">2</td>
<td align="right">1</td>
<td align="right">2</td>
<td align="right">1</td>
<td align="right">2</td>
<td align="right">0</td>
<td align="right">0</td>
</tr>
<tr class="even">
<td align="left">94</td>
<td align="right">0</td>
<td align="right">1</td>
<td align="right">2</td>
<td align="right">1</td>
<td align="right">2</td>
<td align="right">1</td>
<td align="right">1</td>
<td align="right">2</td>
<td align="right">1</td>
<td align="right">2</td>
<td align="right">0</td>
<td align="right">0</td>
</tr>
<tr class="odd">
<td align="left">95</td>
<td align="right">0</td>
<td align="right">0</td>
<td align="right">2</td>
<td align="right">0</td>
<td align="right">2</td>
<td align="right">1</td>
<td align="right">1</td>
<td align="right">2</td>
<td align="right">0</td>
<td align="right">1</td>
<td align="right">0</td>
<td align="right">0</td>
</tr>
<tr class="even">
<td align="left">96</td>
<td align="right">0</td>
<td align="right">1</td>
<td align="right">2</td>
<td align="right">1</td>
<td align="right">2</td>
<td align="right">1</td>
<td align="right">1</td>
<td align="right">2</td>
<td align="right">1</td>
<td align="right">2</td>
<td align="right">0</td>
<td align="right">0</td>
</tr>
<tr class="odd">
<td align="left">97</td>
<td align="right">0</td>
<td align="right">1</td>
<td align="right">1</td>
<td align="right">0</td>
<td align="right">1</td>
<td align="right">1</td>
<td align="right">2</td>
<td align="right">2</td>
<td align="right">1</td>
<td align="right">2</td>
<td align="right">0</td>
<td align="right">0</td>
</tr>
<tr class="even">
<td align="left">98</td>
<td align="right">0</td>
<td align="right">0</td>
<td align="right">2</td>
<td align="right">0</td>
<td align="right">2</td>
<td align="right">0</td>
<td align="right">1</td>
<td align="right">2</td>
<td align="right">0</td>
<td align="right">1</td>
<td align="right">0</td>
<td align="right">0</td>
</tr>
<tr class="odd">
<td align="left">99</td>
<td align="right">0</td>
<td align="right">1</td>
<td align="right">1</td>
<td align="right">0</td>
<td align="right">1</td>
<td align="right">1</td>
<td align="right">2</td>
<td align="right">2</td>
<td align="right">0</td>
<td align="right">2</td>
<td align="right">0</td>
<td align="right">0</td>
</tr>
<tr class="even">
<td align="left">100</td>
<td align="right">0</td>
<td align="right">0</td>
<td align="right">1</td>
<td align="right">0</td>
<td align="right">2</td>
<td align="right">0</td>
<td align="right">1</td>
<td align="right">2</td>
<td align="right">1</td>
<td align="right">2</td>
<td align="right">0</td>
<td align="right">0</td>
</tr>
<tr class="odd">
<td align="left">191</td>
<td align="right">0</td>
<td align="right">0</td>
<td align="right">2</td>
<td align="right">0</td>
<td align="right">0</td>
<td align="right">0</td>
<td align="right">2</td>
<td align="right">1</td>
<td align="right">1</td>
<td align="right">1</td>
<td align="right">0</td>
<td align="right">1</td>
</tr>
<tr class="even">
<td align="left">192</td>
<td align="right">0</td>
<td align="right">0</td>
<td align="right">2</td>
<td align="right">0</td>
<td align="right">0</td>
<td align="right">0</td>
<td align="right">2</td>
<td align="right">1</td>
<td align="right">1</td>
<td align="right">1</td>
<td align="right">0</td>
<td align="right">1</td>
</tr>
<tr class="odd">
<td align="left">193</td>
<td align="right">0</td>
<td align="right">1</td>
<td align="right">2</td>
<td align="right">1</td>
<td align="right">1</td>
<td align="right">1</td>
<td align="right">2</td>
<td align="right">1</td>
<td align="right">1</td>
<td align="right">0</td>
<td align="right">0</td>
<td align="right">1</td>
</tr>
<tr class="even">
<td align="left">194</td>
<td align="right">0</td>
<td align="right">1</td>
<td align="right">2</td>
<td align="right">1</td>
<td align="right">1</td>
<td align="right">1</td>
<td align="right">2</td>
<td align="right">1</td>
<td align="right">1</td>
<td align="right">0</td>
<td align="right">0</td>
<td align="right">1</td>
</tr>
<tr class="odd">
<td align="left">195</td>
<td align="right">0</td>
<td align="right">0</td>
<td align="right">2</td>
<td align="right">0</td>
<td align="right">0</td>
<td align="right">0</td>
<td align="right">2</td>
<td align="right">2</td>
<td align="right">0</td>
<td align="right">2</td>
<td align="right">0</td>
<td align="right">0</td>
</tr>
<tr class="even">
<td align="left">196</td>
<td align="right">0</td>
<td align="right">1</td>
<td align="right">2</td>
<td align="right">1</td>
<td align="right">1</td>
<td align="right">1</td>
<td align="right">2</td>
<td align="right">0</td>
<td align="right">2</td>
<td align="right">0</td>
<td align="right">0</td>
<td align="right">2</td>
</tr>
<tr class="odd">
<td align="left">197</td>
<td align="right">0</td>
<td align="right">0</td>
<td align="right">2</td>
<td align="right">0</td>
<td align="right">0</td>
<td align="right">0</td>
<td align="right">2</td>
<td align="right">0</td>
<td align="right">2</td>
<td align="right">0</td>
<td align="right">0</td>
<td align="right">2</td>
</tr>
<tr class="even">
<td align="left">198</td>
<td align="right">0</td>
<td align="right">0</td>
<td align="right">2</td>
<td align="right">0</td>
<td align="right">0</td>
<td align="right">0</td>
<td align="right">2</td>
<td align="right">1</td>
<td align="right">1</td>
<td align="right">1</td>
<td align="right">0</td>
<td align="right">1</td>
</tr>
<tr class="odd">
<td align="left">199</td>
<td align="right">0</td>
<td align="right">1</td>
<td align="right">2</td>
<td align="right">1</td>
<td align="right">1</td>
<td align="right">1</td>
<td align="right">2</td>
<td align="right">1</td>
<td align="right">1</td>
<td align="right">1</td>
<td align="right">0</td>
<td align="right">1</td>
</tr>
<tr class="even">
<td align="left">200</td>
<td align="right">0</td>
<td align="right">1</td>
<td align="right">2</td>
<td align="right">1</td>
<td align="right">1</td>
<td align="right">1</td>
<td align="right">2</td>
<td align="right">1</td>
<td align="right">1</td>
<td align="right">1</td>
<td align="right">0</td>
<td align="right">1</td>
</tr>
</tbody>
</table>
</div>
<div id="pca-analysis" class="section level2">
<h2><span class="header-section-number">2.4</span> PCA analysis</h2>
<p>The fish x SNP data matrix can be analyzed directly using the function <code>prcomp()</code>. When performing PCA with numeric variables such as gene expression it is common to normalize them by adding <code>scale = TRUE</code>. Despite being in numerical form, the measures of genotype from the SNP are really categorical. Nonetheless, for the purpose of PCA they exist as integers already scaled from 0 to 2. For this reason, the default <code>scale = FALSE</code> is used.</p>
<p>The function <code>prcomp()</code> produces a list of 5 output categories, of which <code>$x</code> and <code>$sdev</code> are used in the downstream analysis here. <code>$x</code> is a data matrix of <em>m</em> rows and <em>n</em> columns. Rows correspond to the number of SNP such that <em>m</em> = 96 observations exist. Columns correspond to principle components which although derived through analysis also corresponds to the number of SNP such that <em>n</em> = 96 variables.</p>
<div class="sourceCode" id="cb5"><pre class="sourceCode r"><code class="sourceCode r"><span id="cb5-1"><a href="differentiation-of-populations-using-pca.html#cb5-1"></a>pca &lt;-<span class="st"> </span><span class="kw">prcomp</span>(snp_pops)</span>
<span id="cb5-2"><a href="differentiation-of-populations-using-pca.html#cb5-2"></a></span>
<span id="cb5-3"><a href="differentiation-of-populations-using-pca.html#cb5-3"></a></span>
<span id="cb5-4"><a href="differentiation-of-populations-using-pca.html#cb5-4"></a>pca.data &lt;-<span class="st"> </span><span class="kw">data.frame</span>(<span class="dt">fish_ID =</span> <span class="kw">rownames</span>(pca<span class="op">$</span>x), <span class="co"># vector of fish IDs</span></span>
<span id="cb5-5"><a href="differentiation-of-populations-using-pca.html#cb5-5"></a>                       <span class="dt">hatchery =</span> <span class="kw">c</span>(<span class="kw">rep</span>(<span class="st">&quot;A&quot;</span>, gen),(<span class="kw">rep</span>(<span class="st">&quot;B&quot;</span>, gen))),</span>
<span id="cb5-6"><a href="differentiation-of-populations-using-pca.html#cb5-6"></a>                       <span class="dt">X =</span> pca<span class="op">$</span>x[,<span class="dv">1</span>], <span class="co"># vector of PC1 in order of fish IDs</span></span>
<span id="cb5-7"><a href="differentiation-of-populations-using-pca.html#cb5-7"></a>                       <span class="dt">Y =</span> pca<span class="op">$</span>x[,<span class="dv">2</span>]) <span class="co"># vector of PC2 in order of fish IDs</span></span></code></pre></div>
</div>
<div id="assessment-of-the-principle-components" class="section level2">
<h2><span class="header-section-number">2.5</span> Assessment of the principle components</h2>
<p>The following code compiles a data frame of principle component effects. A vector of principle component IDs is first made as <code>p_comp</code>. The number of principle components is always equal to the number of subjects, 20 fish in our case. The vector was made simply as 1 to subject number using number of columns from the <code>pca$x</code> matrix. Although column number was used because each column represents one principle component in <code>pca$x</code>, row number which equates to subjects could just as easily be used. After the principle component IDs were called, the standard deviation of the principle components was called as <code>pca$sd</code>. Squaring the standard deviations resulted in <code>pca_var</code>, the variance of the principle components. Due to the ‘additive’ nature of variance, the percent of the total variance could then be calculated for each of the principle components. Finally, the individual vectors were assembled into a data frame and the first few lines shown.</p>
<div class="sourceCode" id="cb6"><pre class="sourceCode r"><code class="sourceCode r"><span id="cb6-1"><a href="differentiation-of-populations-using-pca.html#cb6-1"></a>p_comp &lt;-<span class="st"> </span><span class="dv">1</span><span class="op">:</span><span class="kw">dim</span>(pca<span class="op">$</span>x)[<span class="dv">2</span>] <span class="co"># create a vector pf principle component IDs</span></span>
<span id="cb6-2"><a href="differentiation-of-populations-using-pca.html#cb6-2"></a>pca_sd &lt;-<span class="st"> </span>pca<span class="op">$</span>sdev <span class="co"># call the standard deviation from the PCA results</span></span>
<span id="cb6-3"><a href="differentiation-of-populations-using-pca.html#cb6-3"></a>pca_var &lt;-<span class="st"> </span>pca_sd<span class="op">^</span><span class="dv">2</span> <span class="co"># calculate the variance of the principle components</span></span>
<span id="cb6-4"><a href="differentiation-of-populations-using-pca.html#cb6-4"></a>pca_var_per &lt;-<span class="st"> </span><span class="kw">round</span>(pca_var <span class="op">/</span><span class="st"> </span><span class="kw">sum</span>(pca_var) <span class="op">*</span><span class="st"> </span><span class="dv">100</span>, <span class="dv">1</span>) <span class="co"># calculate the percent variance of the total variance for each principle component</span></span>
<span id="cb6-5"><a href="differentiation-of-populations-using-pca.html#cb6-5"></a></span>
<span id="cb6-6"><a href="differentiation-of-populations-using-pca.html#cb6-6"></a>df_p_comp &lt;-<span class="st"> </span><span class="kw">data.frame</span>(p_comp,</span>
<span id="cb6-7"><a href="differentiation-of-populations-using-pca.html#cb6-7"></a>                        pca_sd,</span>
<span id="cb6-8"><a href="differentiation-of-populations-using-pca.html#cb6-8"></a>                        pca_var,</span>
<span id="cb6-9"><a href="differentiation-of-populations-using-pca.html#cb6-9"></a>                        pca_var_per)</span>
<span id="cb6-10"><a href="differentiation-of-populations-using-pca.html#cb6-10"></a><span class="kw">kable</span>(<span class="kw">head</span>(df_p_comp))</span></code></pre></div>
<table>
<thead>
<tr class="header">
<th align="right">p_comp</th>
<th align="right">pca_sd</th>
<th align="right">pca_var</th>
<th align="right">pca_var_per</th>
</tr>
</thead>
<tbody>
<tr class="odd">
<td align="right">1</td>
<td align="right">4.473026</td>
<td align="right">20.007965</td>
<td align="right">43.5</td>
</tr>
<tr class="even">
<td align="right">2</td>
<td align="right">2.030738</td>
<td align="right">4.123899</td>
<td align="right">9.0</td>
</tr>
<tr class="odd">
<td align="right">3</td>
<td align="right">1.854764</td>
<td align="right">3.440148</td>
<td align="right">7.5</td>
</tr>
<tr class="even">
<td align="right">4</td>
<td align="right">1.719828</td>
<td align="right">2.957808</td>
<td align="right">6.4</td>
</tr>
<tr class="odd">
<td align="right">5</td>
<td align="right">1.693977</td>
<td align="right">2.869558</td>
<td align="right">6.2</td>
</tr>
<tr class="even">
<td align="right">6</td>
<td align="right">1.359560</td>
<td align="right">1.848404</td>
<td align="right">4.0</td>
</tr>
</tbody>
</table>
<p>It is a good idea to look at all principle components together in a figure. This can easily be performed through a <strong>scree plot</strong>. Scree plots are generally presented as line or bar graphs in the form of percent variation as a function of principle component. As per the analysis, the principle components on the x-axis are ordered from the most-to-least influential.</p>
<div class="sourceCode" id="cb7"><pre class="sourceCode r"><code class="sourceCode r"><span id="cb7-1"><a href="differentiation-of-populations-using-pca.html#cb7-1"></a>df_p_comp <span class="op">%&gt;%</span></span>
<span id="cb7-2"><a href="differentiation-of-populations-using-pca.html#cb7-2"></a><span class="st">  </span><span class="kw">ggplot</span>(<span class="kw">aes</span>(<span class="dt">x =</span> p_comp, <span class="dt">y =</span> pca_var_per)) <span class="op">+</span></span>
<span id="cb7-3"><a href="differentiation-of-populations-using-pca.html#cb7-3"></a><span class="st">  </span><span class="kw">geom_point</span>(<span class="dt">stat =</span> <span class="st">&#39;identity&#39;</span>, <span class="dt">color =</span> <span class="st">&quot;blue&quot;</span>, <span class="dt">size =</span> <span class="dv">2</span>) <span class="op">+</span></span>
<span id="cb7-4"><a href="differentiation-of-populations-using-pca.html#cb7-4"></a><span class="st">  </span><span class="kw">geom_line</span>(<span class="dt">linetype =</span> <span class="st">&quot;dotted&quot;</span>, <span class="dt">color =</span> <span class="st">&quot;blue&quot;</span>, <span class="dt">size =</span> <span class="dv">1</span>) <span class="op">+</span></span>
<span id="cb7-5"><a href="differentiation-of-populations-using-pca.html#cb7-5"></a><span class="st">  </span><span class="kw">theme_bw</span>() <span class="op">+</span></span>
<span id="cb7-6"><a href="differentiation-of-populations-using-pca.html#cb7-6"></a><span class="st">  </span><span class="kw">xlab</span>(<span class="st">&quot;principle component&quot;</span>) <span class="op">+</span></span>
<span id="cb7-7"><a href="differentiation-of-populations-using-pca.html#cb7-7"></a><span class="st">  </span><span class="kw">ylab</span>(<span class="st">&quot;effect size (percent)&quot;</span>)</span></code></pre></div>
<p><img src="bookdownproj_files/figure-html/pc_scree-1.png" width="672" /></p>
</div>
<div id="pca-graphical-analysis" class="section level2">
<h2><span class="header-section-number">2.6</span> PCA graphical analysis</h2>
<p>The primary goal of PCA is to reduce the number of independent variables (SNP in our case) needed to differentiate between dependent factors (individuals and their populations in our case)</p>
<p>In the receding figure we see that by generation 10 there is a pretty clear separation of the two hatchery populations. Our next task is to understand the relative effects of the principle components, and then the relative effects of the SNPs within the most influential principle components. In our example we might find that only a few SNP, rather than a low density chip, are all that is needed to differentiate our populations. It would also be a first step in a functional characterization of the genetic differences between the two populations.</p>
<div class="sourceCode" id="cb8"><pre class="sourceCode r"><code class="sourceCode r"><span id="cb8-1"><a href="differentiation-of-populations-using-pca.html#cb8-1"></a>pca.data <span class="op">%&gt;%</span></span>
<span id="cb8-2"><a href="differentiation-of-populations-using-pca.html#cb8-2"></a><span class="st">  </span><span class="kw">ggplot</span>(<span class="kw">aes</span>(<span class="dt">x =</span> X, <span class="dt">y =</span> Y, <span class="dt">label =</span> fish_ID)) <span class="op">+</span></span>
<span id="cb8-3"><a href="differentiation-of-populations-using-pca.html#cb8-3"></a><span class="st">  </span><span class="kw">geom_text</span>(<span class="dt">size =</span> <span class="dv">3</span>, <span class="dt">color =</span> <span class="st">&quot;black&quot;</span>) <span class="op">+</span></span>
<span id="cb8-4"><a href="differentiation-of-populations-using-pca.html#cb8-4"></a><span class="st">  </span><span class="kw">geom_point</span>(<span class="kw">aes</span>(<span class="dt">color =</span> hatchery), <span class="dt">size =</span> <span class="dv">8</span>, <span class="dt">alpha =</span> <span class="fl">0.5</span>) <span class="op">+</span></span>
<span id="cb8-5"><a href="differentiation-of-populations-using-pca.html#cb8-5"></a><span class="st">  </span><span class="kw">xlab</span>(<span class="kw">paste</span>(<span class="st">&quot;PC1 - &quot;</span>, pca_var_per[<span class="dv">1</span>], <span class="st">&quot;%&quot;</span>, <span class="dt">sep=</span><span class="st">&quot;&quot;</span>)) <span class="op">+</span></span>
<span id="cb8-6"><a href="differentiation-of-populations-using-pca.html#cb8-6"></a><span class="st">  </span><span class="kw">ylab</span>(<span class="kw">paste</span>(<span class="st">&quot;PC2 - &quot;</span>, pca_var_per[<span class="dv">2</span>], <span class="st">&quot;%&quot;</span>, <span class="dt">sep=</span><span class="st">&quot;&quot;</span>)) <span class="op">+</span></span>
<span id="cb8-7"><a href="differentiation-of-populations-using-pca.html#cb8-7"></a><span class="st">  </span><span class="kw">scale_color_manual</span>(<span class="dt">values =</span> <span class="kw">c</span>(<span class="st">&quot;A&quot;</span> =<span class="st"> &quot;blue&quot;</span>, <span class="st">&quot;B&quot;</span> =<span class="st"> &quot;grey&quot;</span>)) <span class="op">+</span></span>
<span id="cb8-8"><a href="differentiation-of-populations-using-pca.html#cb8-8"></a><span class="st">  </span><span class="kw">theme_bw</span>() <span class="op">+</span></span>
<span id="cb8-9"><a href="differentiation-of-populations-using-pca.html#cb8-9"></a><span class="st">  </span><span class="kw">ggtitle</span>(<span class="kw">paste</span>(<span class="st">&quot;generation&quot;</span>, c_gen, <span class="dt">sep =</span> <span class="st">&quot; &quot;</span>))</span></code></pre></div>
<p><img src="bookdownproj_files/figure-html/pc_graphical%20analysis-1.png" width="672" /></p>
</div>
<div id="identify-significant-snp" class="section level2">
<h2><span class="header-section-number">2.7</span> Identify significant SNP</h2>

</div>
</div>
            </section>

          </div>
        </div>
      </div>
<a href="index.html" class="navigation navigation-prev " aria-label="Previous page"><i class="fa fa-angle-left"></i></a>
<a href="an-introduction-to-linkage-mapping.html" class="navigation navigation-next " aria-label="Next page"><i class="fa fa-angle-right"></i></a>
    </div>
  </div>
<script src="libs/gitbook-2.6.7/js/app.min.js"></script>
<script src="libs/gitbook-2.6.7/js/lunr.js"></script>
<script src="libs/gitbook-2.6.7/js/clipboard.min.js"></script>
<script src="libs/gitbook-2.6.7/js/plugin-search.js"></script>
<script src="libs/gitbook-2.6.7/js/plugin-sharing.js"></script>
<script src="libs/gitbook-2.6.7/js/plugin-fontsettings.js"></script>
<script src="libs/gitbook-2.6.7/js/plugin-bookdown.js"></script>
<script src="libs/gitbook-2.6.7/js/jquery.highlight.js"></script>
<script src="libs/gitbook-2.6.7/js/plugin-clipboard.js"></script>
<script>
gitbook.require(["gitbook"], function(gitbook) {
gitbook.start({
"sharing": {
"github": false,
"facebook": true,
"twitter": true,
"linkedin": false,
"weibo": false,
"instapaper": false,
"vk": false,
"all": ["facebook", "twitter", "linkedin", "weibo", "instapaper"]
},
"fontsettings": {
"theme": "white",
"family": "sans",
"size": 2
},
"edit": {
"link": "https://github.com/pura-acuicultura/Molecular-methods/edit/master/02-pca.Rmd",
"text": "Edit"
},
"history": {
"link": null,
"text": null
},
"view": {
"link": "https://github.com/pura-acuicultura/Molecular-methods/blob/master/02-pca.Rmd",
"text": null
},
"download": null,
"toc": {
"collapse": "subsection"
}
});
});
</script>

</body>

</html>
